### PR TITLE
feat: Add support for single-node mode without hosts in Clusterfile

### PR DESCRIPTION
**Pull request by LZiHaN**
This pr adds support for running in single-node mode without specifying any hosts in the clusterfile.yaml file. This is useful for users who want to run a Kubernetes cluster on a single machine without having to enter IP addresses.

- [x] support `sealos apply`.
- [x] support `sealos gen`.

refer to #2888

### CODE REVIEW

**Comment by cuisongliu** [comment]
add e2e code for test ?

**Comment by LZiHaN** [comment]
> add e2e code for test ?

sure, I haven't written one before, but I can give it a try.

**Comment by cuisongliu** [comment]
You can refer to how to write images.

**Comment by LZiHaN** [comment]
> You can refer to how to write images.

Cool~ this one？

https://github.com/labring/sealos/blob/e71f980be1d586648b757d6938c8113e7ad1897e/test/e2e/images_test.go#L29

**Comment by cuisongliu** [comment]
> > You can refer to how to write images.
> 
> Cool~ this one？
> 
> https://github.com/labring/sealos/blob/e71f980be1d586648b757d6938c8113e7ad1897e/test/e2e/images_test.go#L29

yes. Then add the e2e workflow script to execute this code

**Comment by LZiHaN** [comment]
> > > You can refer to how to write images.
> > 
> > 
> > Cool~ this one？
> > https://github.com/labring/sealos/blob/e71f980be1d586648b757d6938c8113e7ad1897e/test/e2e/images_test.go#L29
> 
> yes. Then add the e2e workflow script to execute this code

thanks, it helps a lot.

**Comment by LZiHaN** [comment]
hi, @cuisongliu  I am having a workflow problem and I have tried some solutions, but they did not work. can you please help me figure out how to solve this issue?  thank you.

**Comment by cuisongliu** [comment]
Make format

**Comment by cuisongliu** [comment]
You need to add more e2e test cases?

File: pkg/apply/utils.go [hunk 1]
```diff
@@ -135,3 +135,9 @@ func GetNewImages(currentCluster, desiredCluster *v2.Cluster) []string {
 	}
 	return nil
 }
+
+func GetLocalIpv4() string {
```
Comment (cuisongliu) [inline pkg/apply/utils.go:139]: Is it better to move to utilities?

**Comment by LZiHaN** [comment]
> You need to add more e2e test cases?

I can't think of any other test cases at the moment, do you have any suggestions?

File: pkg/apply/utils.go (hunk 1 above)
Comment (LZiHaN) [reply pkg/apply/utils.go:139]: ack

Comment (LZiHaN) [reply pkg/apply/utils.go]: hi, corrected. PTAL

**Comment by cuisongliu** [comment]
> > You need to add more e2e test cases?
> 
> 
> 
> I can't think of any other test cases at the moment, do you have any suggestions?

For example, are the ip generated by apply and gen consistent with the local ip?

**Comment by LZiHaN** [comment]
> > > You need to add more e2e test cases?
> > 
> > 
> > I can't think of any other test cases at the moment, do you have any suggestions?
> 
> For example, are the ip generated by apply and gen consistent with the local ip?

ACK

**Comment by cuisongliu** [comment]
> > > > You need to add more e2e test cases?
> > > 
> > > 
> > > I can't think of any other test cases at the moment, do you have any suggestions?
> > 
> > 
> > For example, are the ip generated by apply and gen consistent with the local ip?
> 
> ACK

It doesn't make much sense to test the cluster image. It is mainly to test the availability of the cluster. You can directly use similar test cases to kubectl get node.

**Comment by LZiHaN** [comment]
> > > > > You need to add more e2e test cases?
> > > > 
> > > > 
> > > > I can't think of any other test cases at the moment, do you have any suggestions?
> > > 
> > > 
> > > For example, are the ip generated by apply and gen consistent with the local ip?
> > 
> > 
> > ACK
> 
> It doesn't make much sense to test the cluster image. It is mainly to test the availability of the cluster. You can directly use similar test cases to kubectl get node.

thanks, very good suggestion, and I can improve that

**Comment by cuisongliu** [comment]
Can you add WeChat to discuss it? My wechat is "cuisongliu".

**Review by cuisongliu — APPROVED** [review]

**Pull request merged**